README.md
[![Test](https://github.com/dietfriends/balance_swiper/actions/workflows/test.yaml/badge.svg)](https://github.com/dietfriends/balance_swiper/actions/workflows/test.yaml)
[![codecov](https://codecov.io/gh/dietfriends/balance_swiper/branch/main/graph/badge.svg?token=pwz0WbL4CL)](https://codecov.io/gh/dietfriends/balance_swiper)
[![pub package](https://img.shields.io/pub/v/balance_swiper.svg)](https://pub.dartlang.org/packages/balance_swiper)

# balance_swiper

The best swiper for flutter , with multiple layouts, infinite loop. Compatible with Android & iOS.


# :sparkles::sparkles: New Features:Layout for pagination.

We are using this project [flutter_page_indicator](https://github.com/best-flutter/flutter_page_indicator) now .

# :sparkles::sparkles: New Features:PageTransformer

Finally, we have `PageTransformer` like android, just set a `transformer` to `Swiper`,
it returns a widget that has been transformed. For now, only support for layout `DEFAULT`.
Thanks to @FlutterRocks ,you've done great job 👏.

We are using this project [transformer_page_view](https://github.com/best-flutter/transformer_page_view) now .


# :sparkles::sparkles: New Features:Layout


![](https://github.com/jzoom/images/raw/master/layout1.gif)

![](https://github.com/jzoom/images/raw/master/layout2.gif)

![](https://github.com/jzoom/images/raw/master/layout3.gif)

[See More](#build-in-layouts)


# Showcases

![Horizontal](https://github.com/dietfriends/balance_swier/raw/master/example/res/1.gif)

![Vertical](https://github.com/dietfriends/balance_swier/raw/master/example/res/2.gif)

![Custom Pagination](https://github.com/dietfriends/balance_swier/raw/master/example/res/3.gif)

![Custom Pagination](https://github.com/dietfriends/balance_swier/raw/master/example/res/4.gif)

![Phone](https://github.com/dietfriends/balance_swier/raw/master/example/res/5.gif)

![Example](https://github.com/jzoom/images/raw/master/swiper-example.gif)

[See More](#codes)

## Roadmap

>see:[ROADMAP.md](https://github.com/dietfriends/balance_swier/blob/master/ROADMAP.md)

## Changelogs

>see:[CHANGELOG.md](https://github.com/dietfriends/balance_swier/blob/master/CHANGELOG.md)

## Getting Started

- [Installation](#installation)
- [Basic Usage](#basic-usage)
- [Constructor](#constructor)
  + [Basic](#basic)
  + [Pagination](#pagination)
  + [Control buttons](#control-buttons)
  + [Controller](#controller)
  + [Autoplay](#autoplay)
- [Build in layouts](#build-in-layouts)
- [Codes](#codes)

### Installation

Add 

```bash

flutter_swiper : ^lastest_version

```
to your pubspec.yaml ,and run 

```bash
flutter packages get 
```
in your project's root directory.

### Basic Usage

Create a new project with command

```
flutter create myapp
```

Edit lib/main.dart like this:

```

import 'package:flutter/material.dart';

import 'package:flutter_swiper/flutter_swiper.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text(widget.title),
      ),
    body:  new Swiper(
        itemBuilder: (BuildContext context,int index){
          return new Image.network("http://via.placeholder.com/350x150",fit: BoxFit.fill,);
        },
        itemCount: 3,
        pagination: new SwiperPagination(),
        control: new SwiperControl(),
      ),
    );
  }
}

```



### Constructor


#### Basic

| Parameter  | Default   | Description |
| :------------ |:---------------:| :-----|
| scrollDirection | Axis.horizontal  | If `Axis.horizontal`, the scroll view's children are arranged horizontally in a row instead of vertically in a column. |
| loop | true |Set to `false` to disable continuous loop mode. |
| index | 0 |  Index number of initial slide. |
| autoplay | false |Set to `true` enable auto play mode. |
| onIndexChanged | void onIndexChanged(int index)  | Called with the new index when the user swiped or autoplay |
| onTap | void onTap(int index)  | Called when user tap ui. |
| duration | 300.0  | The milliscends of every transaction animation costs  |
| pagination | null | set `new SwiperPagination()` to show default pagination
| control | null | set `new SwiperControl()` to show default control buttons


#### Pagination

The pagination extends from `SwiperPlugin`,the `SwiperPlugin` provides extra ui for `Swiper`.Set `new SwiperPagination()` to show default pagination.


| Parameter  | Default   | Description |
| :------------ |:---------------:| :-----|
| alignment | Alignment.bottomCenter  | Change this value if you what to put pagination in other place |
| margin | const EdgeInsets.all(10.0) | The distance between inner side of the parent container. |
| builder | SwiperPagination.dots | There are two default styles `SwiperPagination.dots` and `SwiperPagination.fraction`,both can be customized. |

If you'd like to customize your own pagination, you can do like this:

```
new Swiper(
    ...,
    pagination:new SwiperCustomPagination(
        builder:(BuildContext context, SwiperPluginConfig config){
            return new YourOwnPaginatipon();
        }
    )
);

```



#### Control buttons

The control also extends from `SwiperPlugin`,set `new SwiperControl()` to show default control buttons.


| Parameter  | Default   | Description |
| :------------ |:---------------:| :-----|
| iconPrevious | Icons.arrow_back_ios  | The icon data to display `previous` control button |
| iconNext | Icons.arrow_forward_ios | The icon data to display `next`. |
| color | Theme.of(context).primaryColor | Control button color |
| size | 30.0 | Control button size |
| padding | const EdgeInsets.all(5.0) | Control button padding |


#### Controller

The `Controller` is used to control the `index` of the Swiper, start or stop autoplay.You can create a controller by `new SwiperController()` and save the instance by futher usage.


| Method  | Description |
| :------------ |:-----|
| void move(int index, {bool animation: true}) | Move to the spicified `index`,with animation or not |
| void next({bool animation: true}) | Move to next |
| void previous({bool animation: true}) | Move to previous |
| void startAutoplay() | Start autoplay |
| void stopAutoplay() | Stop autoplay |



#### Autoplay

| Parameter  | Default   | Description |
| :------------ |:---------------:| :-----|
| autoplayDelay | 3000  | Autoplay delay milliseconds. |
| autoplayDisableOnInteraction | true | If set true, `autoplay` is disabled when use swipes. |

## Build in layouts
![](https://github.com/jzoom/images/raw/master/layout1.gif)

```
new Swiper(
  itemBuilder: (BuildContext context, int index) {
    return new Image.network(
      "http://via.placeholder.com/288x188",
      fit: BoxFit.fill,
    );
  },
  itemCount: 10,
  viewportFraction: 0.8,
  scale: 0.9,
)

```



![](https://github.com/jzoom/images/raw/master/layout2.gif)

```
new Swiper(
  itemBuilder: (BuildContext context, int index) {
    return new Image.network(
      "http://via.placeholder.com/288x188",
      fit: BoxFit.fill,
    );
  },
  itemCount: 10,
  itemWidth: 300.0,
  layout: SwiperLayout.STACK,
)
```

![](https://github.com/jzoom/images/raw/master/layout3.gif)

```
new Swiper(
    itemBuilder: (BuildContext context, int index) {
      return new Image.network(
        "http://via.placeholder.com/288x188",
        fit: BoxFit.fill,
      );
    },
    itemCount: 10,
    itemWidth: 300.0,
    itemHeight: 400.0,
    layout: SwiperLayout.TINDER,
 )
```


![](https://github.com/jzoom/images/raw/master/layout4.gif)

Very easy to create you own custom animation:
```

 new Swiper(
  layout: SwiperLayout.CUSTOM,
  customLayoutOption: new CustomLayoutOption(
      startIndex: -1,
      stateCount: 3
  ).addRotate([
    -45.0/180,
    0.0,
    45.0/180
  ]).addTranslate([
    new Offset(-370.0, -40.0),
    new Offset(0.0, 0.0),
    new Offset(370.0, -40.0)
  ]),
  itemWidth: 300.0,
  itemHeight: 200.0,
  itemBuilder: (context, index) {
    return new Container(
      color: Colors.grey,
      child: new Center(
        child: new Text("$index"),
      ),
    );
  },
  itemCount: 10)

```

The `CustomLayoutOption` is designed to describe animations.
It is very easy to specify every state of items in Swiper.

```
new CustomLayoutOption(
      startIndex: -1,  /// Which index is the first item of array below      
      stateCount: 3    /// array length 
  ).addRotate([        // rotation of every item 
    -45.0/180,
    0.0,
    45.0/180
  ]).addTranslate([           /// offset of every item
    new Offset(-370.0, -40.0),
    new Offset(0.0, 0.0),
    new Offset(370.0, -40.0)
  ])

```

## Codes

![Example](https://github.com/jzoom/images/raw/master/swiper-example.gif)

```
new ConstrainedBox(
  child: new Swiper(
    outer:false,
    itemBuilder: (c, i) {
      return new Wrap(
        runSpacing:  6.0,
        children: [0,1,2,3,4,5,6,7,8,9].map((i){
          return new SizedBox(
            width: MediaQuery.of(context).size.width/5,
            child: new Column(
              mainAxisSize: MainAxisSize.min,
              children: <Widget>[
                new SizedBox(
                  child:  new Container(
                    child: new Image.network("https://fuss10.elemecdn.com/c/db/d20d49e5029281b9b73db1c5ec6f9jpeg.jpeg%3FimageMogr/format/webp/thumbnail/!90x90r/gravity/Center/crop/90x90"),
                  ),
                  height: MediaQuery.of(context).size.width * 0.12,
                  width: MediaQuery.of(context).size.width * 0.12,
                ),
                new Padding(padding: new EdgeInsets.only(top:6.0),child: new Text("$i"),)
              ],
            ),
          );
        }).toList(),
      );
    },
    pagination: new SwiperPagination(
      margin: new EdgeInsets.all(5.0)
    ),
    itemCount: 10,
  ),
    constraints:new BoxConstraints.loose(new Size(screenWidth, 170.0))
),

```





You can find all custom options here:

>https://github.com/dietfriends/balance_swier/blob/master/example/lib/src/ExampleCustom.dart







CHANGELOG.md
## [1.0.0] - [2021/05/27]

- migrate to null safety

# Forked from flutter_swiper

## [1.1.5] - [2019/03/10]

 * Fix findRenderObject is null


## [1.1.4] - [2018/10/18]

## [1.1.2] - [2018/10/10]
    * Fix #7
    * Support layout for pagination.

## [1.1.1] - [2018/09/20]
    * Fix `test` failure.

## [1.1.0] - [2018/09/20]
    * Fix index bug ,See #11
    * Enable `autoplayDisableOnInteraction` property, if set `true`,disable autoplay when user swipes.

## [1.0.7] - [2018/09/02]
    * Don't call SwiperController's dispose when `Swiper` dispose.

## [1.0.6] - [2018/08/28]
    * Implement vertical scroll direction for `TINDER` and `STACK` layout, see #9

## [1.0.5] - [2018/08/09]
    * Add feature : support fade for `viewportFraction`

## [1.0.4] - [2018/07/18]
    * Fix some typo,thanks to [csharad](https://github.com/csharad)

## [1.0.3] - [2018/07/18]
    * Use new to create everything. See #5

## [1.0.2] - [2018/07/16]
    * Fix #4

## [1.0.1] - [2018/07/11]
    * Add layout (CUSTOM) so that you can create your own layout

## [1.0.0] - [2018/07/08]
    * Add layouts ( DEFAULT,STACK,TINDER )
    * Allow to put pagination outer of the swiper container.

## [0.0.9] - [2018/06/10]
    * Add ci
    * Add testcase

## [0.0.8] - [2018/06/07]
    * And chinese document
    * Update infinity_page_view version to 1.0.0

## [0.0.7] - [2018/05/24]
    * The default color of pagination is ThemeData.scaffoldBackgroundColor and ThemeData.primaryColor
    * The default color of control buttons is ThemeData.primaryColor and ThemeData.disabledColor
    * The alignment of pagination is Alignment.bottomCenter by default when scrollDirection== Axis.horizontal, Alignment.centerRight by default when scrollDirection== Axis.vertical
    * Change default value of `autoplay` to false

## [0.0.6] - [2018/05/24]
    * Fix index bug
        
## [0.0.5] - [2018/05/24]
    * Fix zero itemCount bug

## [0.0.4] - [2018/05/20]
    * Update README

## [0.0.3] - [2018/05/20]
    * Update README
    * Support none loop mode
    * Add more examples

## [0.0.2] - [2018/05/20]
    * Autoplay
    * Plugins support 
    * Examples

## [0.0.1] - [2018/05/20]
    * Infinite loop
    * Control buttons
    * Pagination
    * Custom control buttons
    * Custom pagination
    * Controler

example/lib/src/ExampleCustom.dart
import 'package:flutter/material.dart';
import 'package:balance_swiper/balance_swiper.dart';
import 'config.dart';
import 'forms/form_widget.dart';

class ExampleCustom extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return _ExampleCustomState();
  }
}

class _ExampleCustomState extends State<ExampleCustom> {
  //properties want to custom
  late int _itemCount;

  late bool _loop;

  late bool _autoplay;

  late int _autoplayDelay;

  late double _padding;

  late bool _outer;

  late double _radius;

  late double _viewportFraction;

  late SwiperLayout _layout;

  late int _currentIndex;

  late double _scale;

  late Axis _scrollDirection;

  late Curve _curve;

  late double _fade;

  late bool _autoplayDisableOnInteraction;

  late CustomLayoutOption customLayoutOption;

  late SwiperController _controller;

  TextEditingController numberController = TextEditingController();

  Widget _buildItem(BuildContext context, int index) {
    return ClipRRect(
      borderRadius: BorderRadius.all(Radius.circular(_radius)),
      child: Image.asset(
        images[index % images.length],
        fit: BoxFit.fill,
      ),
    );
  }

  @override
  void didUpdateWidget(ExampleCustom oldWidget) {
    customLayoutOption = CustomLayoutOption(startIndex: -1, stateCount: 3)
        .addRotate([-45.0 / 180, 0.0, 45.0 / 180]).addTranslate(
            [Offset(-370.0, -40.0), Offset(0.0, 0.0), Offset(370.0, -40.0)]);
    super.didUpdateWidget(oldWidget);
  }

  @override
  void initState() {
    customLayoutOption = CustomLayoutOption(startIndex: -1, stateCount: 3)
        .addRotate([-25.0 / 180, 0.0, 25.0 / 180]).addTranslate(
            [Offset(-350.0, 0.0), Offset(0.0, 0.0), Offset(350.0, 0.0)]);
    _fade = 1.0;
    _currentIndex = 0;
    _curve = Curves.ease;
    _scale = 0.8;
    _controller = SwiperController();
    _layout = SwiperLayout.TINDER;
    _radius = 10.0;
    _padding = 0.0;
    _loop = true;
    _itemCount = 3;
    _autoplay = false;
    _autoplayDelay = 3000;
    _viewportFraction = 0.8;
    _outer = false;
    _scrollDirection = Axis.horizontal;
    _autoplayDisableOnInteraction = false;
    super.initState();
  }

// maintain the index

  Widget buildSwiper() {
    return Swiper(
      onTap: (int index) {
        Navigator.of(context)
            .push(MaterialPageRoute(builder: (BuildContext context) {
          return Scaffold(
            appBar: AppBar(
              title: Text('New page'),
            ),
            body: Container(),
          );
        }));
      },
      customLayoutOption: customLayoutOption,
      fade: _fade,
      index: _currentIndex,
      onIndexChanged: (int index) {
        setState(() {
          _currentIndex = index;
        });
      },
      curve: _curve,
      scale: _scale,
      itemWidth: 300.0,
      controller: _controller,
      layout: _layout,
      outer: _outer,
      itemHeight: 200.0,
      viewportFraction: _viewportFraction,
      autoplayDelay: _autoplayDelay,
      loop: _loop,
      autoplay: _autoplay,
      itemBuilder: _buildItem,
      itemCount: _itemCount,
      scrollDirection: _scrollDirection,
      indicatorLayout: PageIndicatorLayout.COLOR,
      autoplayDisableOnInteraction: _autoplayDisableOnInteraction,
      pagination: SwiperPagination(
          builder: const DotSwiperPaginationBuilder(
              size: 20.0, activeSize: 20.0, space: 10.0)),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Column(children: <Widget>[
      Container(
        color: Colors.black87,
        child: SizedBox(
            height: 300.0, width: double.infinity, child: buildSwiper()),
      ),
      Expanded(
          child: ListView(
        children: <Widget>[
          Text('Index:$_currentIndex'),
          Row(
            children: <Widget>[
              ElevatedButton(
                onPressed: () {
                  _controller.previous(animation: true);
                },
                child: Text('Prev'),
              ),
              ElevatedButton(
                onPressed: () {
                  _controller.next(animation: true);
                },
                child: Text('Next'),
              ),
              Expanded(
                  child: TextField(
                controller: numberController,
              )),
              ElevatedButton(
                onPressed: () {
                  var text = numberController.text;
                  setState(() {
                    _currentIndex = int.parse(text);
                  });
                },
                child: Text('Update'),
              ),
            ],
          ),
          FormWidget(
            label: 'layout',
            child: FormSelect(
              placeholder: 'Select layout',
              value: _layout,
              values: [
                SwiperLayout.DEFAULT,
                SwiperLayout.STACK,
                SwiperLayout.TINDER,
                SwiperLayout.CUSTOM
              ],
              valueChanged: (SwiperLayout value) {
                _layout = value;
                setState(() {});
              },
            ),
          ),
          FormWidget(
            label: 'scrollDirection',
            child: Switch(
                value: _scrollDirection == Axis.horizontal,
                onChanged: (bool value) => setState(() => _scrollDirection =
                    value ? Axis.horizontal : Axis.vertical)),
          ),
          FormWidget(
            label: 'autoplayDisableOnInteraction',
            child: Switch(
                value: _autoplayDisableOnInteraction,
                onChanged: (bool value) =>
                    setState(() => _autoplayDisableOnInteraction = value)),
          ),
          //Pannel Begin
          FormWidget(
            label: 'loop',
            child: Switch(
                value: _loop,
                onChanged: (bool value) => setState(() => _loop = value)),
          ),
          FormWidget(
            label: 'outer',
            child: Switch(
                value: _outer,
                onChanged: (bool value) => setState(() => _outer = value)),
          ),
          //Pannel Begin
          FormWidget(
            label: 'autoplay',
            child: Switch(
                value: _autoplay,
                onChanged: (bool value) => setState(() => _autoplay = value)),
          ),

          FormWidget(
            label: 'padding',
            child: NumberPad(
              number: _padding,
              step: 5.0,
              min: 0.0,
              max: 30.0,
              onChangeValue: (num value) {
                _padding = value.toDouble();
                setState(() {});
              },
            ),
          ),
          FormWidget(
            label: 'scale',
            child: NumberPad(
              number: _scale,
              step: 0.1,
              min: 0.0,
              max: 1.0,
              onChangeValue: (num value) {
                _scale = value.toDouble();
                setState(() {});
              },
            ),
          ),
          FormWidget(
            label: 'fade',
            child: NumberPad(
              number: _fade,
              step: 0.1,
              min: 0.0,
              max: 1.0,
              onChangeValue: (num value) {
                _fade = value.toDouble();
                setState(() {});
              },
            ),
          ),
          FormWidget(
            label: 'itemCount',
            child: NumberPad(
              number: _itemCount,
              step: 1,
              min: 0,
              max: 100,
              onChangeValue: (num value) {
                _itemCount = value.toInt();
                setState(() {});
              },
            ),
          ),

          FormWidget(
            label: 'radius',
            child: NumberPad(
              number: _radius,
              step: 1.0,
              min: 0.0,
              max: 30.0,
              onChangeValue: (num value) {
                this._radius = value.toDouble();
                setState(() {});
              },
            ),
          ),

          FormWidget(
            label: 'viewportFraction',
            child: NumberPad(
              number: _viewportFraction,
              step: 0.1,
              max: 1.0,
              min: 0.5,
              onChangeValue: (num value) {
                _viewportFraction = value.toDouble();
                setState(() {});
              },
            ),
          ),

          FormWidget(
            label: 'curve',
            child: FormSelect(
              placeholder: 'Select curve',
              value: _layout,
              values: [
                Curves.easeInOut,
                Curves.ease,
                Curves.bounceInOut,
                Curves.bounceOut,
                Curves.bounceIn,
                Curves.fastOutSlowIn
              ],
              valueChanged: (Curve value) {
                _curve = value;
                setState(() {});
              },
            ),
          ),
        ],
      ))
    ]);
  }
}
